### ISSUE TITLE
Lag when running louvre-views and weston clone

### ISSUE THREAD

**Issue by renhiyama** [CuarzoSoftware/Louvre#75]
Moving apps on screen makes it lag, sometimes my cursor feels laggy (stops moving for a sec, etc). Disabling vsync creates a lot of screen tearing. Running btop in terminal doesnt show any spike in gpu usage when moving apps. Apps work flawlessly on Gnome.

Running fastfetch under any of these two examples will show there's two displays: one is the native display that is shown on both gnome and under the examples. However, I see another display when running inside example, it says: 
```
Display (DRM connector): 2560x1900 @ 60Hz
```

I am not sure whether this is intended way, but my laptop screen size is 1920x1200 @ 60Hz, which is shown properly on the other display. I believe this new display might be the source of all the problems, or atleast I wonder something is wrong with this - considering it uses higher display resolution that my monitor allows.

**Comment by ehopperdietzel** [comment]
Interesting, are you using the latest versions of Louvre and SRM? Also, could you provide the output of running `drm_info`?

**Comment by renhiyama** [comment]
> Interesting, are you using the latest versions of Louvre and SRM? Also, could you provide the output of running `drm_info`?

I am running louvre from fedora copr repo. Here's the link: https://copr.fedorainfracloud.org/coprs/ngompa/louvre/

It seems like the build is one month old. Let me know if should I uninstall it and compile from source?

**Comment by renhiyama** [comment]
> A laggy cursor suggests it might be due to the new DRM atomic API. Could you try running it with `SRM_RENDER_MODE_ITSELF_FB_COUNT=2 SRM_FORCE_LEGACY_API=1 louvre-views`?

Currently I'm in class, I'll look onto it once I reach my hostel. Btw, could I know where should I learn this drm stuff, and other compositing stuff, so I could work on this project?

**Comment by ehopperdietzel** [comment]
> I am running louvre from fedora copr repo. Here's the link: https://copr.fedorainfracloud.org/coprs/ngompa/louvre/

That version should be ok.

> Currently I'm in class, I'll look onto it once I reach my hostel. Btw, could I know where should I learn this drm stuff, and other compositing stuff, so I could work on this project?

Sure, in Louvre and in Wayland compositors in general the following is used:

- **GBM**: For allocating GPU buffers with various formats, modifiers, and usages, such as textures, rendering, or scanout. Modifiers define the memory layout (e.g., linear, tiled, etc.), with each GPU vendor typically supporting their own specific modifiers, though nearly all support the linear modifier. This becomes particularly important in multi-GPU setups where buffers need to be shared between GPUs.

- **DRM/KMS**: Provides control over displays, such as setting gamma, controlling vsync, etc. For example, it allows you to display GBM buffers on screen.

- **EGL**: Acts as middleware between a "native" platform and OpenGL. In the context of a Wayland compositor, it allows you to create an EGLSurface (used for OpenGL rendering) from a GBM surface or an EGLImage from a GBM buffer or other sources, which can then be used as OpenGL textures. In this case, the platform would be GBM, but it could also be Wayland, X11, etc., if you were developing a Wayland or X11 OpenGL application, respectively.

- **OpenGL**: You may already know, for accelerated rendering.

Instead of dealing with DRM/KMS/EGL/GBM directly, Louvre has a DRM backend that relies on [SRM](https://github.com/CuarzoSoftware/SRM). Therefore, the cursor issue is related to SRM.

DRM/KMS is quite complex to learn, and there isn't much documentation available, especially from a userspace perspective. I learned from various sources, including the kernel documentation, the Mesa source code, and examples like kmscube.

There is also a guy named Simon Ser who has some excellent introductory resources:

https://www.youtube.com/watch?v=haes4_Xnc5Q
https://github.com/ascent12/drm_doc/blob/master/01_intro/01_intro.md

I hope this helps! And feel free to ask me.

**Comment by renhiyama** [comment]
Wow! Thanks for the huge amount of awesome info @ehopperdietzel ♥️

Btw, I just realised using louvre just broke gnome dark theming. 
![image](https://github.com/user-attachments/assets/67c74275-d695-4b22-ba26-dacd1d121ffa)

I hope you know how to fix this :skull: I could accept this problem if it only happened on louvre-views or some other example, but this is occuring on gnome desktop, which means louvre has messed up stuff somewhere :(

**Comment by renhiyama** [comment]
Btw, just ran those two commands that you asked me to:
`drm_info` : 
[drm.txt](https://github.com/user-attachments/files/16868679/drm.txt)

And running `SRM_RENDER_MODE_ITSELF_FB_COUNT=2 SRM_FORCE_LEGACY_API=1 louvre-views` fixes the cursor laggy issue! Moving windows seems to be fine too :+1:

**Comment by ehopperdietzel** [comment]
> Btw, I just realized using Louvre just broke GNOME dark theming.

Louvre and the examples don't handle changing toolkit styles. I believe GNOME uses D-Bus to request apps to switch their themes at runtime. However, all the examples attempt to restart the XDG desktop portal and start `xdg-desktop-portal-wlr` during startup, mainly to enable apps to capture the screen. Commenting out this line in `louvre-views` might resolve the issue (only for `louvre-views`):

[examples/louvre-views/src/main.cpp line 25](https://github.com/CuarzoSoftware/Louvre/blob/2ad1cf77db6a7c51cf2090e024ef1f58eca1cccb/src/examples/louvre-views/src/main.cpp#L25).

> Btw, do you have any suggestions on how to run it "natively" without messing up my native filesystem?

Louvre and the examples require certain assets to be in specific locations. If you don’t want to install it on your system, you could specify a custom `--prefix` when building with Meson. This way, when you run `meson install`, it installs everything in a specific folder.

> I don't want to switch TTY when developing Louvre examples.

If you're running a Wayland DE, you can launch the examples as Wayland clients. This should happen automatically, but if not, ensure the `WAYLAND_DISPLAY` environment variable is set and matches the one of your DE.

> Btw, I just ran those two commands that you asked me to:  
`drm_info`:  
[drm.txt](https://github.com/user-attachments/files/16868679/drm.txt)

Interesting, it shows two connected displays: one is your laptop screen (eDP), and the other is a virtual display of type "writeback." I'm not sure what this is used for, but it could be the cause of the slowdown. I'll blacklist connectors with that type, which should fix it.

> And running `SRM_RENDER_MODE_ITSELF_FB_COUNT=2 SRM_FORCE_LEGACY_API=1 louvre-views` fixes the cursor lag issue! Moving windows seems to be fine too. 👍

Great! Thanks for testing. Could you also check if the issue is fixed when running it only with `SRM_RENDER_MODE_ITSELF_FB_COUNT=2 louvre-views`?

**Comment by renhiyama** [comment]
> Great! Thanks for testing. Could you also check if the issue is fixed when running it only with SRM_RENDER_MODE_ITSELF_FB_COUNT=2 louvre-views?

The issue comes back with this. 

> Interesting, it shows two connected displays: one is your laptop screen (eDP), and the other is a virtual display of type "writeback." I'm not sure what this is used for, but it could be the cause of the slowdown. I'll blacklist connectors with that type, which should fix it.

I assumed so!

> If you're running a Wayland DE, you can launch the examples as Wayland clients. This should happen automatically, but if not, ensure the WAYLAND_DISPLAY environment variable is set and matches the one of your DE.

![image](https://github.com/user-attachments/assets/5ccb93e8-03a0-4a1f-8fcd-b55c3a64f107)

It doesnt create a whole screen dedicated to louvre-views, worse still - I cant even take my cursor near the end of that new window to show the dock, so I cant start nautilus or any app from inside it :(
Btw do you know how to run louvre-views in fullscreen mode if that's the case?

> Louvre and the examples require certain assets to be in specific locations. If you don’t want to install it on your system, you could specify a custom --prefix when building with Meson. This way, when you run meson install, it installs everything in a specific folder.

Thanks for this info. I was trying out running from source (compiling myself) via a fedora container inside distrobox, but apparently vscode intellisense isn't intelligent enough to figure where to read header files from :laughing:


I will try to learn more about louvre and make one of my example soon too :)

**Comment by renhiyama** [comment]
Btw it would be faster for both of us to have chat available via discord, what do you think? my id is: `renhiyama.gg` :smile:

**Comment by ehopperdietzel** [comment]
> Btw do you know how to run louvre-views in fullscreen mode if that's the case?

There is a shortcut option in GNOME settings to toggle fullscreen mode, but it's unassigned by default. You can also make the dock appear by dragging the cursor down.

> It might be faster for both of us to chat via Discord. What do you think? My ID is renhiyama.gg 😄

We have a [Matrix group](https://cuarzosoftware.github.io/Louvre/contact_page.html) maybe you could join us there instead :) 

I believe I've fixed the issues. I disabled the [WRITEBACK_CONNECTORS](https://dri.freedesktop.org/docs/drm/gpu/drm-uapi.html#c.DRM_CLIENT_CAP_WRITEBACK_CONNECTORS) DRM client capability in SRM, which should prevent that additional display from appearing. I think those "virtual displays" are used to access the pixel data resulting from DRM planes composition, meaning there might be a GPU-to-main-memory copy involved, which could explain the slowdown.

Unfortunately, I can't test if that fixes the cursor bugs. Could you try the [disable_writeback](https://github.com/CuarzoSoftware/SRM/tree/disable_writeback) branch of SRM and run the following commands from a TTY?

```
SRM_FORCE_LEGACY_API=0 louvre-views
SRM_FORCE_LEGACY_API=1 louvre-views
```

Hopefully, the cursor issue won't persist in either case.

**Comment by ehopperdietzel** [comment]
If you’re new to using Meson and Git, here are some instructions:
First, remove the Louvre Fedora package to prevent any conflicts. Then, follow these steps:

### Installing SRM

```bash
$ git clone https://github.com/CuarzoSoftware/SRM.git
$ cd SRM
$ git checkout disable_writeback
$ cd src
$ meson setup builddir
$ cd builddir
$ meson install
$ sudo ldconfig
```
You can find the dependencies [here](https://cuarzosoftware.github.io/SRM/md_md__downloads.html). Keep in mind that this process will install the library on your system. When running `meson install`, all installed files will be listed, allowing you to remove them later if needed.

### Installing Louvre

Next, you may need to recompile Louvre using a similar approach:

```bash
$ git clone https://github.com/CuarzoSoftware/Louvre.git
$ cd Louvre
$ git checkout zoom
$ cd src
$ meson setup builddir
$ cd builddir
$ meson install
$ sudo ldconfig
```
You can find the dependencies [here](https://cuarzosoftware.github.io/Louvre/downloads_page.html).

If you encounter linking errors when launching the compositors, ensure that you set the environment variable:

```bash
# From a free tty
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/path/to/srm_and_louvre_libs

# Run with the atomic DRM API (I am especially interested in whether the cursor works correctly in this case)
SRM_FORCE_LEGACY_API=0 louvre-views

# Run with the legacy DRM API (should still work fine)
SRM_FORCE_LEGACY_API=1 louvre-views
```

**Comment by ehopperdietzel** [comment]
This should now be fixed with SRM v0.7.2.
